olive/engine/footprint.py

```diff
@@ -0,0 +1,178 @@
+from datetime import datetime
+from pathlib import Path
+
+from olive.common.config_utils import ConfigBase, config_json_dumps, config_json_loads
+
+logger = logging.getLogger(__name__)
+
+
+class FootprintNodeMetric(ConfigBase):
+    value: dict
+    cmp_direction: dict = {}
+    is_goals_met: bool = False
+
+
+class FootprintNode(ConfigBase):
+    # None for no parent which means current model is the input model
+    parent_model_id: str = None
+    model_id: str
+    model_config: dict = None
+    from_pass: str = None
+    pass_run_config: dict = None
+    is_pareto_frontier: bool = False
+    metrics: FootprintNodeMetric = FootprintNodeMetric(value={}, is_goals_met=False)
+
+    date_time: float = datetime.now().timestamp()
+
+    def update(self, **kwargs):
+        for k, v in kwargs.items():
+            setattr(self, k, v)
+
+
+class Footprint:
+    """
+    The engine footprint is a class that contains the footprint of the engine runtime.
+    It is used to collect the runtime state of the Olive engine and to organize the state
+    in a way that is easy to visualize and understand for customers.
+    """
+
+    def __init__(
+        self,
+        nodes: OrderedDict = {},
+        objective_dict: dict = {},
+        is_marked_pareto_frontier: bool = False,
+    ):
+        self.nodes = nodes
+        self.objective_dict = objective_dict
+        self.is_marked_pareto_frontier = is_marked_pareto_frontier
+
+    def record_objective_dict(self, objective_dict):
+        self.objective_dict = objective_dict
+
+    def resolve_metrics(self):
+        for k, v in self.nodes.items():
+            if v.metrics is None:
+                continue
+            is_goals_met = []
+            for metric_name in v.metrics.value:
+                if metric_name in self.objective_dict:
+                    cmp_direction = 1 if self.objective_dict[metric_name]["higher_is_better"] else -1
+                    self.nodes[k].metrics.cmp_direction[metric_name] = cmp_direction
+                    _goal = self.objective_dict[metric_name]["goal"]
+                    if _goal is None:
+                        is_goals_met.append(True)
+                    else:
+                        is_goals_met.append(v.metrics.value[metric_name] * cmp_direction >= _goal)
+                else:
+                    logger.warning(f"Metric {metric_name} is not in the objective dict")
+            self.nodes[k].metrics.is_goals_met = all(is_goals_met)
+
+    def record(self, foot_print_node: FootprintNode = None, **kwargs):
+        _model_id = kwargs.get("model_id", None)
+        if foot_print_node is not None:
+            _model_id = foot_print_node.model_id
+            self.nodes[_model_id] = foot_print_node
+        elif _model_id in self.nodes:
+            self.nodes[_model_id].update(**kwargs)
+        else:
+            self.nodes[_model_id] = FootprintNode(**kwargs)
+        self.resolve_metrics()
+
+    def get_candidates(self):
```

Comment:
When we evaluate the baseline, the input model also gets evaluated. So does it also become a candidate?

Reply:
Also, what about the models created by intermediate passes in `pass-by-pass` execution order? They also get evaluated. 

Maybe we need a `is_candiate` flag for models that is true if the pass is the final pass? In the future we could also allow intermediate passes to also be candidates by using a similar flag when registering the pass.

Reply:
Remove the input model from candidates.

Yes, I saw your previous draft PR. There is a flag called is_candidate for final around model. But I am thinking that:
if the intermediate model(in pass-by-pass) is in pareto-frontier, we may should put it in output to let user choose. So I just skip to add is_candidate.

Reply:
That also sounds good to me. Thanks for the clarification. I was wondering about cases where the intermediate passes might not have the model in the desired format (like pytorch model -> pytorch model pass, when we want onnx model) but we don't have such cases currently.